### PR TITLE
Identity perf tests

### PR DESCRIPTION
**Pull request by gearama**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/externals/perf.combo.pem [hunk 1]
```diff
@@ -0,0 +1,53 @@
+Bag Attributes
```
Comment (RickWinter) [inline sdk/identity/azure-identity/test/perf/inc/azure/identity/test/externals/perf.combo.pem:1]: Is this pem file going to trigger security scanner failures?  Could we store this in keyvault and download it as part of the test prep?

Comment (LarryOsterman) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/externals/perf.combo.pem:1]: This is 100% going to trigger credscan.

Comment (LarryOsterman) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/externals/perf.combo.pem:1]: If it doesn't, something is wrong in how credscan is configured :).

Comment (gearama) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/externals/perf.combo.pem:1]: updated credscan file to allow this cert. having this cert in keyvault does not work, i tried it as keyvault will download only cert backups not certs themselves in the right format

Comment (LarryOsterman) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/externals/perf.combo.pem:1]: How do we ensure that nobody can use this certificate to access any azure services? Is there a way to say "Allow logon but don't allow any service access"?

**Review by antkmsft — COMMENTED** [review]
The code looks good to me, and the cert situation is also not completely clear to me, let's get some meeting to discuss with the team?

**Comment by gearama** [comment]
That is how the object constructor is defined, how the sample is written and how we expect customers to use it, by providing a certificate path. 
What would be your suggestion to test this scenario, while still  trying to mimic how customers are using it? 
If we want to use test recordings we need to have a deterministic certificate we cannot generate it on the fly. I checked with engineering and they have no issue with checking in this cert which is self generated by me for this specific test case. 
Or are you saying we are not allowed to test this class because we don't want cert files in the repo?

**Comment by antkmsft** [comment]
@gearama, are any of the deployed services in our subscription accept this certificate?
I.e. can a random guy grab it, and authenticate to use our test resources in Azure?

**Comment by antkmsft** [comment]
Overall, it might be good to have one test certificate checked in, as long as it is not uploaded to Azure. We kind of have one in the tests (https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/identity/azure-identity/test/ut/client_certificate_credential_test.cpp#L373-L493), but it not used for live tests, only for the tests that run locally. You can even run that test with no network connection, and how about this one - does it make actual network requests to Azure services?

**Comment by gearama** [comment]
no , this certificate is not used for nay live services, it is used just to generate the recordings that i will add later on. when the creds are initialized i need to pass in a valid cert file that gets validated during the construction of the instance. thus the need for a valid one.

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/client_certificate_credential_test.hpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    /**
+     * @brief Define the test
+     *
+     * @param context The cancellation token.
+     */
+    void Run(Azure::Core::Context const& context) override
+    {
+      auto t = m_credential->GetToken(m_tokenRequestContext, context);
+    }
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/perf/inc/azure/identity/test/client_certificate_credential_test.hpp:70]: How do we run this test locally (what commands do we use) and can you please share some perf test run results?

How's the overhead here that we are measuring?

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#include <string>
+#include <vector>
+
+namespace Azure { namespace Identity { namespace Test {
+
+  /**
+   * @brief A test to measure the authentication token performance.
+   *
+   */
+  class EnvironmentCredentialTest : public Azure::Perf::PerfTest {
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp:25]: Is the goal here to compare the overhead of different `TokenCredentials` within C++ or compare it across languages?

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+     * @brief Setup the test
+     *
+     */
+    void Setup() override
+    {
+      m_tokenRequestContext.Scopes.push_back(m_options.GetMandatoryOption<std::string>("Scope"));
+      if (!m_options.GetOptionOrDefault<bool>("Cache", false))
+      {
+        m_tokenRequestContext.MinimumExpiration = std::chrono::hours(1000000);
+      }
```
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp:41]: What's the intent here? Why do we want to set this `MinimumExpiration`?

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/client_certificate_credential_test.hpp (hunk 2 above)
Comment (ahsonkhan) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/client_certificate_credential_test.hpp:70]: For example, do we pass in the path to the checked-in `perf.combo.pem` file as a `CertPath` option when running this test?

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp (hunk 3 above)
Comment (gearama) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp:25]: both

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp (hunk 4 above)
Comment (gearama) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp:41]: to not use the credentials cache

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/client_certificate_credential_test.hpp (hunk 2 above)
Comment (gearama) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/client_certificate_credential_test.hpp:70]: the command is documented in the Get test option method

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp (hunk 3 above)
Comment (ahsonkhan) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp:25]: And what are the results/findings from that comparison?

**Review by ahsonkhan — APPROVED** [review]
I would love to see the results of the run here and what our takeaways are. If you ran them locally and have some data, please do share.

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp (hunk 3 above)
Comment (gearama) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/environment_credential_test.hpp:25]: well , for the moment we have no history to compare results to just the "previous version" that we choose based on pipeline settings.

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/client_certificate_credential_test.hpp (hunk 2 above)
Comment (LarryOsterman) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/client_certificate_credential_test.hpp:70]: For information on how to run performance tests, see [the documentation](https://github.com/Azure/azure-sdk-for-cpp/blob/azure-core_1.11.2/doc/PerformanceTesting.md).

File: sdk/identity/azure-identity/test/perf/inc/azure/identity/test/externals/perf.combo.pem (hunk 1 above)
Comment (LarryOsterman) [reply sdk/identity/azure-identity/test/perf/inc/azure/identity/test/externals/perf.combo.pem:1]: Or is this a bogus certificate that doesn't grant access to any resources?

File: eng/CredScanSuppression.json [hunk 5]
```diff
@@ -8,6 +8,10 @@
     {
       "file": [ "sdk/attestation/azure-security-attestation/test/ut/crypto_test_collateral.cpp" ],
       "_justification": "File contains a PEM encoded RSA private key used by test code."
+    },
+    {
+      "file": [ "sdk/identity/azure-identity/test/perf/inc/azure/identity/test/externals/perf.combo.pem" ],
+      "_justification": "File contains a PEM encoded RSA private key used by perf test code."
```
Comment (LarryOsterman) [inline eng/CredScanSuppression.json:14]: What is that private key used for? How do we know that it's not the key to some existing Azure resource?

Comment (gearama) [reply eng/CredScanSuppression.json:14]: No , this certificate is not used for no live services, it is used just to generate the recordings that i will add later on. when the creds are initialized i need to pass in a valid cert file that gets validated during the construction of the instance. thus the need for a valid one.

Comment (LarryOsterman) [reply eng/CredScanSuppression.json:14]: To be clear: I'm looking for some kind of documentation in the code which will explain why this private key isn't a security vulnerability.

Right now, all that I see is that the certificate is used in authentication, and that's very concerning.

**Pull request merged**